### ISSUE TITLE
using external buffer possible?

### ISSUE THREAD

**Issue by ericlangedijk** [rust-windowing/softbuffer#222]
Currently I am copying my own rgba into the buffer of the softbuffer-surface.
I was wondering if that could be faster. And if it would be possible that the softbuffer uses this external buffer as source.
I saw that in Windows the drawing is done with BitBlt (which I believe is hardware-accelerated) and also was wondering if StretchDIBits would be an option. The last one does not need a source-handle, but don't know if StretchDIBits is as optimized.

**Comment by ids1024** [comment]
Could your program be refactored to write into the `softbuffer::Buffer` instead of a separate buffer?

At least on some backends, that should perform better than any possible implementation that presents from just a `&[u8]` or `&[u32]`. Not sure about the performance of `StretchDIBits`, `BitBlt`, and such. (Other than that SDL uses `BitBlt` with a comment that when implemented it ended up being faster than trying to use DirectX.)

**Comment by ericlangedijk** [comment]
Yes BitBlt is fast.
I was thinking about your suggestion as well.
I could probably rewrite my home-made Bitmap (with all kinds of draw functions) to write to a 'backend' instead of its own Vec<Rgba. Lot of work but doable.
It saves a copy of some megabytes each frame....

**Comment by notgull** [comment]
Many drawing interfaces have a way to render to a mutable byte slice. For instance in my own code I use [`tiny-skia::PixmapMut`](https://docs.rs/tiny-skia/latest/tiny_skia/struct.PixmapMut.html) around the softbuffer buffer in order to render in software to the  window.

**Comment by ericlangedijk** [comment]
Yes I think for certain backends it will not be possible or handy. No clue about other systems.
On windows we have a u32 array which happily has exactly the same layout as my inner bitmap on which i perform all my (cpu) drawing logic buffer scaling and rendering.

**Comment by daxpedda** [comment]
Maybe some platform-specific extensions would be more appropriate then.

**Comment by ids1024** [comment]
If you want to write something specifically for Windows and have some very specific requirements, you can also always just do that directly with windows-rs instead of using Softbuffer.

I guess it's an open design question in softbuffer to decide when to offer platform extensions vs suggest using platform APIs directly.

**Comment by notgull** [comment]
> I guess it's an open design question in softbuffer to decide when to offer platform extensions vs suggest using platform APIs directly.

I would say this is definitely a case where platform extensions should not be used. The changes required for some platforms to use direct buffers would effectively require re-architecting the entire crate.

**Comment by ericlangedijk** [comment]
A little update on this: I managed to refactor my drawing. Enabling a 'borrowed bitmap', borrowing the softbuffer pixels.
The result are quite ok! On a WIndows PC fullscreen (3840 x 2160) I get around 500 frames per second (including all game logic).
I scale my inner game bitmap (320x200) up to the softbuffer pixels (3200x2000) and then call the 'present' function. Great!

**Comment by MarijnS95** [comment]
On Windows, I was told that there are a variety of importing functions that allow (specially allocated?) memory ranges to be imported into all kinds of objects.

Anyway, the point was to import/store `&[u32]` into some form of `Buffer` abstraction that - in the current `softbuffer` implementation - already needs to be copied to an implementation-specific format, for example on Android which currently stores a `Vec<u32>` for this purpose.

And it won't be necessary when `softbuffer` finally manages to zero-copy map the underlying abstraction on these platforms (with diverging pixel formats and strides) too.

---

> Yeah, though I'm unsure how it'd look API-wise, and it'd make it harder for us to do double-buffering internally if found to be desired.

Unsure if we're talking about the same kind of double-buffering. In my mind, the user allocates 2 buffers (or we cycle them when returning to the user), the user draws into it, and we ultimately (zero-copy!) present it to the display while the user could **concurrently** lock and draw into another buffer.

Afaik that's what Android's surface (un)lock function "should" be doing under the hood, but it's better exposed via `ASurfaceControl` + `AHardwareBuffer` (an interface I tried wrapping into `softbuffer` on a private branch, hope I can finish that some day).

**Comment by madsmtm** [comment]
I think I understood, and I think we're in agreement. I'll close this as "not planned", with the intent to work towards correctly using memory-mapped zero-copy buffers instead.
